File: queue/queue_ll.cpp
/* Queue DSA implementation using linked lists in C++ witn OOP */
#include <iostream>

/* Node struct */
struct Node
{
public:
    /* Data members */
    int data;
    Node *next;

    /* Default constructor */
    Node()
        : data(), next(NULL)
    {
    }
};

/* Queue class */
class Queue
{
    /* Data members */
private:
    Node *front, *rear;
    /* Default constructor - empty queue */
public:
    Queue()
        : front(NULL), rear(NULL)
    {
    }
    /* Public member functions */

    // Create a new node
    Node *createNewNode(int data)
    {
        // Dynamic memory allocation on the heap for a new node
        Node *newNode = new Node();
        // Assign values
        newNode->data = data;
        // Point the next pointer to null since we're going to be push a new node at the back of the queue
        newNode->next = NULL;
        return newNode;
    }

    // Enqueue
    void enqueue(int data)
    {
        // Create a new node for the queue
        Node *newNode = createNewNode(data);
        // Check if the queue is empty
        if (!front)
        {
            front = rear = newNode;
            return;
        }
        // Point the last node to the new node
        rear->next = newNode;
        // Shift the rear pointer to the new node
        rear = newNode;
    }

    // Dequeue
    void dequeue()
    {
        // Check if the queue is empty
        if (!front)
        {
            std::cout << "The queue is empty!";
            return;
        }
        // Assign a temp variable to the front of the queue
        Node *temp = front;
        // Shift the front pointer to the next node
        front = front->next;
        // If front is greater than rear, reset the queue
        if (front > rear)
        {
            front = rear = NULL;
        }
        // Deallocate memory for the temp variable to pop out the first node
        delete temp;
    }

    // Print queue
    void printQueue()
    {
        // Traverse the queue from the start and print the value of each node on each iteration
        Node *temp = front;
        while (temp)
        {
            std::cout << temp->data << " ";
            temp = temp->next;
        }
        std::cout << std::endl;
    }
};

int main()
{
    // Instantiate the queue class (create an empty queue)
    Queue q;
    q.enqueue(5);
    q.enqueue(10);
    q.enqueue(25);
    q.printQueue();
    q.dequeue();
    q.dequeue();
    q.printQueue();

    return 0;
}

File: queue/queue_ll.py
# Queue DSA in Python using liked lists

# Node class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Queue class


class Queue:
    def __init__(self):
        self.front = None
        self.rear = None

    # Methods

    # Enqueue
    def enqueue(self, data):
        # Create a new node
        new_node = Node(data)
        new_node.next = None
        # Check if the queue is empty
        if (not self.front):
            # Assign the new node to the front and the back of the queue
            self.front = self.rear = new_node
            return
        # Point the rear node to the new node
        self.rear.next = new_node
        # Shift the rear pointer to the new node
        self.rear = new_node

    # Dequeue
    def dequeue(self):
        # Check if the queue is empty, return
        if (not self.front):
            print('The queue is empty!')
            return
        # Shift the front to the next node and free memory allocated for the previous node
        temp = self.front
        self.front = self.front.next
        temp = None
        # If the front is greater than the rear, reset the queue
        if (self.front == None):
            self.rear = None

    # Print the queue
    def print_queue(self):
        # Set a temp variable to traverse the queue (linked list) starting at the head
        temp = self.front
        while (temp):
            print(temp.data, end=' ')
            temp = temp.next
        print()


if __name__ == '__main__':
    # Instantiate the queue class
    q = Queue()
    q.enqueue(5)
    q.enqueue(10)
    q.enqueue(20)
    q.print_queue()
    q.dequeue()
    q.dequeue()
    q.print_queue()
